rfd/0085-improve-webapps-codebase.md

```diff
@@ -0,0 +1,163 @@
+---
+authors: Ryan Clark (ryan.clark@goteleport.com)
+state: draft
+---
+
+# RFD 85 - Improving the Webapps codebase
+
+## What
+
+Improve the webapps codebase by changing the way we test our components and write our component stories.
+
+## Why
+
+The current convention is to use Storybook to render our large, full components in their different states.  This has a 
+couple of knock-on effects.
+
+### Component Architecture
+
+As it's hard to mock data in Storybook, we end up exporting a component that is suitable for rendering in Storybook,
+such as:
+
+```typescript jsx
+export function DesktopSession(props: State) {
+  const {
+    directorySharingState,
+    setDirectorySharingState,
+    clipboardState,
+    fetchAttempt,
+    tdpConnection,
+    disconnected,
+    wsConnection,
+    setTdpConnection,
+  } = props;
+  // etc
+  
+  return (
+    <Flex flexDirection="column">
+      // etc
+    </Flex>
+  );
+}
+```
+
+This component takes all of its state and business logic from the props passed in. 
+
+For use in the actual application, we wrap the component above in a default export which populates the props the 
+component needs by using a hook.
+
+```typescript jsx
+export default function Container() {
+  const state = useDesktopSession();
+  return <DesktopSession {...state} />;
+}
+```
+
+
+### Testing
+
+As a result of the above, we end up testing the `DesktopSession` component instead of the actual component (`Container`) that is
+rendered in the UI. 
+
+A story may look like this:
+
+```typescript jsx
+export const FetchError = () => (
+  <DesktopSession
+    {...props}
+    fetchAttempt={{ status: 'failed', statusText: 'some fetch  error' }}
+    tdpConnection={{ status: 'success' }}
+    wsConnection={'open'}
+    disconnected={false}
+  />
+);
+```
+
+Which is then tested like so:
+
+```typescript jsx
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the UI part of the application,
+but skipping all of the business logic.
+
+Storybook isn't best suited for the way we are using it. As a result of testing our large components primarily via stories, we have
+ended up only testing our UI and nothing else.
+
+## Details
+
+### Changing how we write stories
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components. We should therefore restrict usage of Storybook mostly to
+our shared component library, and test our fullscreen components another way.
```

Comment:
As for testing complex components through writing stories and then running snapshot tests on them – I completely agree that it might not be the best way to go about it. Some of those components might deserve more traditional tests.

However, I still think Storybook is immensely useful for some of the bigger, one-time-use components. When I was adding a new feature to `DocumentGateway`, the stories I wrote were super useful. I was able to make a small change which impacted the component in a couple of different states and then I could verify visually that everything still looks okay.

![DocumentGateway stories](https://user-images.githubusercontent.com/27113/186648165-dad7e7e3-010a-4f3f-bf70-619032690dad.png)

In this way, I was using Storybook more to do a visual inspection rather than testing. It also serves as a great documentation – the next dev that has to modify that component can see what are the possible states the components can be in and which were important enough that someone made a story for them.

Just as a couple of you, I wasn't ever big on snapshot tests, I don't think writing testing components through snapshots of stories is the way to go. But I don't think it warrants removing stories for more complex components altogether nor changing the approach with "dumb/smart" components.

---

Though one problem with "dumb/smart" components to address might be the way we pass props around which indeed kinda obscures how a particular function/value gets to a component.

```typescript jsx
  const state = useDesktopSession();
  return <DesktopSession {...state} />;
```

Reply:
i'm +1 with @ravicious, writing stories also helped me catch a lot of bugs by visually checking out all possible states (most recent experience was with `FormLogin` which had so many possible state combos), i also rely on storybook to quickly demo different states, or quickly look up a state that people had questions on, and also as a way to quickly review our current design and copy (i once went through our entire storybook with our copy writer).

Reply:
I think it makes sense to reword this so that the core message is that we're moving away from using Storybook as the primary testing methodology, but still leave room for visual testing.

@JanKaczmarkiewicz also brought up the possibility of using [Chromatic](https://www.chromatic.com/) for testing, which seems like it gives us the benefits of automated visual comparisons without the drawbacks of snapshots.

Reply:
I've reworded it a bit to say that writing stories is still okay, but we shouldn't be basing our tests off of them.

Reply:
Sounds good for me now!